include/swift/Parse/PersistentParserState.h

```diff
@@ -81,11 +84,34 @@ class PersistentParserState {
     {}
   };
 
+  class DelayedDeclListState {
+    friend class PersistentParserState;
+    friend class Parser;
+    unsigned Flags;
+    DeclContext *ParentContext;
+    ParserPos BodyPos;
+    SourceLoc BodyEnd;
+    SavedScope Scope;
+
+    SavedScope takeScope() {
+      return std::move(Scope);
+    }
+
+  public:
+    DelayedDeclListState(unsigned Flags,
+                     DeclContext *ParentContext, SourceRange BodyRange,
+                     SourceLoc PreviousLoc, SavedScope &&Scope)
+    : Flags(Flags), ParentContext(ParentContext), BodyPos{BodyRange.Start, PreviousLoc},
+      BodyEnd(BodyRange.End), Scope(std::move(Scope))
+    {}
+  };
+
   bool InPoundLineEnvironment = false;
   // FIXME: When condition evaluation moves to a later phase, remove this bit
   // and adjust the client call 'performParseOnly'.
   bool PerformConditionEvaluation = true;
 private:
+  ASTContext &Ctx;
```

Comment:
You can always recover the `ASTContext` from an `IterableDeclContext`.

Reply:
this is also for removing lazyParser from the `ASTContext` when `PersistentParserState` dies. We cannot ensure there're `IterableDeclContext` available for us to recover `ASTContext`.